packages/expo-updates/android/src/main/java/expo/modules/updates/statemachine/UpdatesStateMachine.kt

```diff
@@ -0,0 +1,221 @@
+package expo.modules.updates.statemachine
+
+import android.content.Context
+import com.facebook.react.bridge.Arguments
+import com.facebook.react.bridge.WritableMap
+import expo.modules.updates.logging.UpdatesLogger
+import org.json.JSONObject
+
+/**
+ * The Updates state machine class. There should be only one instance of this class
+ * in a production app, instantiated as a property of UpdatesController.
+ */
+class UpdatesStateMachine constructor(
+  androidContext: Context? = null
+) {
+
+  /**
+   * The context and logger are null in unit tests,
+   * but are instantiated in a production app.
+   */
+  private val logger = when (androidContext) {
+    null -> null
+    else -> UpdatesLogger(androidContext)
+  }
+
+  var changeEventSender: UpdatesStateChangeEventSender? = null
+
+  /**
+   * The current state
+   */
+  var state: UpdatesStateValue = UpdatesStateValue.Idle
+
+  /**
+   * The context
+   */
+  var context: UpdatesStateContext = UpdatesStateContext()
+
+  /**
+   Called after the app restarts (reloadAsync()) to reset the machine to its
+   starting state.
+   */
+  fun reset() {
+    state = UpdatesStateValue.Idle
+    context = UpdatesStateContext()
+    logger?.info("Updates state change: reset, context = ${context.json}")
+    sendChangeEventToJS()
+  }
+
+  /**
+   Called by LoaderTask delegate methods in UpdatesController during the initial
+   background check for updates, and called by checkForUpdateAsync(), fetchUpdateAsync(), and reloadAsync().
+   */
+  fun processEvent(event: UpdatesStateEvent) {
+    if (transition(event)) {
+      context = reducedContext(context, event)
+      logger?.info("Updates state change: ${event.type}, context = ${context.json}")
+      // Send change event
+      sendChangeEventToJS(event)
+    }
+  }
+
+  /**
+   Make sure the state transition is allowed, and then update the state.
+   */
+  private fun transition(event: UpdatesStateEvent): Boolean {
+    val allowedEvents: List<UpdatesStateEventType> = updatesStateAllowedEvents[state] ?: listOf()
+    if (!allowedEvents.contains(event.type)) {
+      // Optionally put an assert here when testing to catch bad state transitions in E2E tests
```

Comment:
For the JS API calls, we should not see any invalid state transitions, since they are queued in AsyncTask on Android and a dispatch queue on iOS. For the check and download that happens on launch, it's a bit more difficult, but we should definitely give some thought to how we can guarantee that only valid transitions will occur.

I did test this by adding an assert that would fire whenever an invalid transition is attempted, and crash the app. No such crashes are occurring in any of the E2E tests, so I have some confidence that they won't happen. However, we can add some chaos testing in E2E (press check button multiple times, etc.) to test this more thoroughly.

Reply:
I added some tests in the state machine E2E where the check button is pressed multiple times in a row, then download pressed multiple times. Logs show that state transitions are occurring as expected. I'll add some more testing like this.

Reply:
I've added a new project in Linear for future work on the state machine and JS API, and added this and other tasks. https://linear.app/expo/project/expo-updates-js-api-2-de38dafeb7e4/ENG

Reply:
ENG-8998